### ISSUE TITLE
Inconsistent residual Gibbs energy for PR EoS and reference state not working

### ISSUE THREAD

**Issue by dallan-keylogic** [CoolProp/CoolProp#2068]
### Description

The residual chemical potential and fugacity coefficient satisfy the relationship mu_j^res = RT log(phi_j). The residual Gibbs energy can be calculated by the sum of chemical potentials weighted by mole fractions. However, the residual Gibbs energy calculated by CoolProp in this way using a cubic EoS does not agree with what the `gibbsmolar_residual` function returns. 

I encountered this error when trying to [check the implementation of a cubic EoS in IDAES](https://idaes-pse.readthedocs.io/en/stable/) with that of CoolProp. The residual enthalpy calculations matched, but the residual entropy from `smolar_residual` did not. However, when I worked backwards to calculate the residual entropy from the residual enthalpy and the Gibbs energy calculated by use of fugacity coefficients using CoolProp, I was able to obtain agreement with what IDAES calculated. As a result, I believe the error is in the residual entropy calculations.

One final issue---apparently setting the reference state for cubics does not work.

### Steps to Reproduce

```
from math import log
import CoolProp.CoolProp as CP

AS = CP.AbstractState("PR","H2&N2&NH3")

Ntot = 18 + 53 + 385
mole_fracs = [53/Ntot,18/Ntot,385/Ntot]
AS.set_mole_fractions(mole_fracs)
AS.specify_phase(CP.iphase_gas)
AS.update(CP.PT_INPUTS,4E7,573)

print("Molar residual Gibbs energy")
print(AS.gibbsmolar_residual())
print(8.314472*573*sum(xj*log(AS.fugacity_coefficient(j)) 
                       for j,xj in enumerate(mole_fracs)))

print("")
print("Failure to change reference state")
print(CP.PropsSI("Hmolar","P",1E5,"T",298,"PR::Hydrogen"))
Dmolar = CP.PropsSI("Dmolar", "P", 1E5, "T", 298, "PR::Hydrogen")
CP.set_reference_state("Hydrogen",298,Dmolar,0,0)
print(CP.PropsSI("Hmolar","P",1E5,"T",298,"PR::Hydrogen"))
```

**Expected behavior:** Both calculations of the molar residual Gibbs energy should produce the same number and changing the reference state should zero the enthalpy of Hydrogen at the conditions.

**Actual behavior:** Two dramatically different numbers of the residual Gibbs energy are produced, and the enthalpy returned does not change after a new reference state is set.

### Versions

**CoolProp Version:** 6.4.1
**Operating System and Version:** Windows 10, Version 10.0.19043 Build 19043

**Access Method:** Python through pip

**Comment by ibell** [comment]
Can you please provide what you believe to be the correct values? Also
please specify precisely how you have defined the residual properties.
There are at least three different definitions I know of. I have a hunch
you are mixing residual property definitions.

**Comment by dallan-keylogic** [comment]
In this case, I believe the correct value for the residual Gibbs energy to be the -953.257 J/mol that the second calculation produces.

It's entirely possible that I'm mixing up definitions, or at least not using the definition in the same way that CoolProp is. I believe it to be the value of a property as calculated by the full EoS minus the value of the property as calculated for an ideal gas. In our implementation in IDAES, we're calculating the ideal gas term using the Shomate equation with coefficients from NIST.

**Comment by ibell** [comment]
Ideal gas at same temperature and pressure or at same temperature and
density?

**Comment by dallan-keylogic** [comment]
I was using the same temperature and pressure definition, but I'm guessing CoolProp uses the same temperature and density definition. The latter doesn't make a tremendous amount of sense for Cubic EoSes, but it's probably to be consistent with what's used in the other methods of calculation CoolProp provides. That addresses the first point until further notice.

That leaves the issue with the reference state, unless I'm misunderstanding how setting reference states works.

**Comment by ibell** [comment]
Yes,  all reference EOS are Helmholtz energy explicit, with temperature and
density as independent variables. So too SAFT EoS, etc. Cubics are
relatively unique having a natural dependence on pressure. So make sure you
always define what you mean by residual properties. Otherwise, lots of
opportunities for confusion.

I'm surprised you don't get an error from your reference state setting.
 Want to fix it in CoolProp?

Also,  you might be interested in my work project:
https://github.com/usnistgov/teqp .  Directly related, and forthcoming
paper covering the details.

**Comment by ibell** [comment]
Closing as part of an issue-tracker cleanup — this issue has had no activity in over two years.

*To be transparent: this comment and the close action are being performed by Claude (Anthropic's AI coding assistant), running on behalf of @ibell. Ian reviewed and approved the bulk-close criteria; I am the one executing the action.*

If you would like this issue reconsidered, please feel free to **re-open it** — a fresh comment with current behaviour against the latest CoolProp release is the most useful way to revive the discussion.

— Claude (acting for @ibell)
